### PR TITLE
feat(lib): allow chaining of TerraformIterator created resources

### LINKED ISSUE #1940: Support passing a primitive value (e.g. a string) to `iterator.dynamic()`
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

https://github.com/hashicorp/terraform-cdk/pull/1882 introduced block level iterators with a design that can also be passed to properties that are lists and not blocks. Currently we allow passing an object (which enables users to create arrays of objects) only. But we could also allow e.g. strings, which would enable the following:
```ts
new ArchiveProvider.DataArchiveFile(this, "dir", {
      type: "zip",
      outputPath: "${path.module}/out1.zip",
      sourceDir: files.path,
      excludes: it.dynamic(`${it.value}.txt`), // <- this is not possible yet but would be neat
});
```

One downside to this would be the type handling through JSII, as `object | string` would appear as `interface{}` (aka `any`) in languages like e.g. Golang.

### References

* https://github.com/hashicorp/terraform-cdk/blob/4da8eaeccb62023806cfbbad0a4af868fd5d26db/test/typescript/iterators/main.ts#L34

### LINKED ISSUE #2024: Support accessing resources created by an iterator
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Iterators allow you to create a dynamic amount of resources based on data only know at Terraform execution time; however, if you then need to reference those resources for some other purpose, there isn't a supported way (besides escape hatches). This is still dynamic information so not everything will be possible, but some operations should still work. We could support creating another iterator from the resource which could then be feed to another resource. We may also be able to expose list/map operations for direct access (via Terraform lookups).

### References

- https://discuss.hashicorp.com/t/how-to-iterate-over-objects-made-with-an-iterator/43115
- https://github.com/hashicorp/terraform-cdk/issues/1094

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
This PR brings both chaining iterators to the table as well as a mechanism to craft for expressions from iterators.

Closes #2024 
Closes #1940

### CODE REVIEW

File: packages/cdktf/lib/terraform-iterator.ts [hunk 1]
```diff
@@ -80,6 +81,26 @@ export abstract class TerraformIterator implements ITerraformIterator {
     return new MapTerraformIterator(map);
   }
 
+  /**
+   * Creates a new iterator from a resource that
+   * has been created with the `for_each` argument.
+   */
+  public static fromResources(
```
Comment (ansgarm) [inline packages/cdktf/lib/terraform-iterator.ts:88]: This method is named in plural while the next is singular – I don't have any opinion but I'd want consistency 😄

Comment (DanielMSchmidt) [reply packages/cdktf/lib/terraform-iterator.ts:88]: Should be plural on both indicating we are talking about lists of resources. Where lists means they were created through for each

File: packages/cdktf/lib/terraform-iterator.ts [hunk 2]
```diff
@@ -80,6 +81,26 @@ export abstract class TerraformIterator implements ITerraformIterator {
...
+  ): ResourceTerraformIterator {
+    return new ResourceTerraformIterator(resource);
+  }
+
+  /**
+   * Creates a new iterator from a data source that
+   * has been created with the `for_each` argument.
+   */
+  public static fromDataSources(
+    resource: ITerraformResource
```
Comment (ansgarm) [inline packages/cdktf/lib/terraform-iterator.ts:99]: Is there a specific type for DataSources? Or is the method just doubled for keeping the name simple?

File: packages/cdktf/lib/terraform-iterator.ts [hunk 3]
```diff
@@ -284,3 +305,47 @@ export class MapTerraformIterator extends TerraformIterator {
...
+      throw new Error(
+        "Cannot create iterator from resource without for_each argument"
+      );
+    }
+  }
+
+  /**
+   * Returns the currenty entry in the list or set that is being iterated over.
+   * For lists this is the same as `iterator.value`. If you need the index,
+   * use count using the escape hatch:
```
Comment (ansgarm) [inline packages/cdktf/lib/terraform-iterator.ts:331]: > using the escape hatch

Don't we already have support for `TerraformCount` nowadays?

File: website/docs/cdktf/concepts/iterators.mdx [hunk 4]
```diff
@@ -434,6 +434,212 @@ new TeamMembers(this, "members", {
 
 </CodeTabs>
 
+## Chaining Iterators
+
+Sometimes the need arises to loop over resources created through an iterator; commonly referrred as chaining.
```
Comment (ansgarm) [inline website/docs/cdktf/concepts/iterators.mdx:439]:
```suggestion
Sometimes the need arises to loop over resources created through an iterator; commonly referred to as chaining.
```

**Review by ansgarm — COMMENTED** [review]
Looks good so far, only a few small things 👍

File: packages/cdktf/lib/terraform-iterator.ts (hunk 3 above)
Comment (DanielMSchmidt) [reply packages/cdktf/lib/terraform-iterator.ts:331]: Yeah, I'll fix the comment from where I copied it over from as well 👍

File: packages/cdktf/lib/terraform-iterator.ts (hunk 2 above)
Comment (DanielMSchmidt) [reply packages/cdktf/lib/terraform-iterator.ts:99]: It's just an alias so that we don't get people confused on if it's only supported for Resources :)

File: packages/cdktf/lib/terraform-iterator.ts [hunk 5]
```diff
@@ -222,6 +251,88 @@ export abstract class TerraformIterator implements ITerraformIterator {
...
     );
   }
+
+  /**
+   * Creates a for expression that maps the iterators to its keys.
+   * For lists these would be the indices, for maps the keys.
+   * As this returns an IResolvable you might need to wrap the output in
+   * a Token, e.g. `Token.asString`.
+   */
+  public mapToKey(): IResolvable {
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf/lib/terraform-iterator.ts:261]: That said, `mapToValuePropertyList` would feel quite off, so I'm unsure here 😅 Still not entirely happy with the name, but not a blocker

File: packages/cdktf/lib/tfExpression.ts [hunk 6]
```diff
@@ -234,8 +263,12 @@ class PropertyAccess extends TFExpression {
 }
 
 // eslint-disable-next-line jsdoc/require-jsdoc
-export function propertyAccess(target: Expression, args: Expression[]) {
-  return new PropertyAccess(target, args) as IResolvable;
+export function propertyAccess(
+  target: Expression,
+  args: Expression[],
+  forceNoBraces = false
```
Comment (ansgarm) [inline packages/cdktf/lib/tfExpression.ts:269]: I don't see any code setting this to true – is this change still required? It seems like you solved the problem with the `resolveExpressionPart` change?

File: packages/cdktf/lib/tfExpression.ts [hunk 7]
```diff
@@ -131,6 +151,11 @@ export function rawString(str: string): IResolvable {
   return new RawString(str);
 }
 
+// eslint-disable-next-line jsdoc/require-jsdoc
+export function unescapedString(str: string): IResolvable {
+  return new UnescapedString(str);
+}
```
Comment (ansgarm) [inline packages/cdktf/lib/tfExpression.ts:157]: This function and the class don't seem to be called / documented or tested anywhere? Is this still required?

File: packages/cdktf/lib/terraform-iterator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf/lib/terraform-iterator.ts:261]: Maybe `mapToKey` could become `keys`. I'm not sure where to go with the others, I see that map is used in multiple contexts, but map is also the most common word for the kind of transformation it does

[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf/lib/terraform-iterator.ts]: I renamed `mapToKeys` and `mapToValues` to `keys` and `values` and for `mapToValuePropertyList` I went into underscore.js which has the `pluck` method that does the same so I renamed it to that

File: packages/cdktf/lib/tfExpression.ts (hunk 6 above)
Comment (DanielMSchmidt) [reply packages/cdktf/lib/tfExpression.ts]: Yeah that was old, removed it.

File: packages/cdktf/lib/terraform-iterator.ts [hunk 8]
```diff
@@ -222,6 +251,88 @@ export abstract class TerraformIterator implements ITerraformIterator {
...
+        this._getForEachExpression(),
+        propertyAccess(FOR_EXPRESSION_VALUE, [property])
+      )
+    );
+  }
+
+  /**
+   * Creates a for expression that results in a list.
+   * This method allows you to create every possible for expression, but requires more knowledge about
+   * Terraforms for expression syntax.
```
Comment (mutahhir) [inline packages/cdktf/lib/terraform-iterator.ts:297]:
```suggestion
   * Terraform's for expression syntax.
```

File: packages/cdktf/lib/terraform-iterator.ts [hunk 9]
```diff
@@ -222,6 +251,88 @@ export abstract class TerraformIterator implements ITerraformIterator {
...
+   * For the most common use cases you can use keys(), values(), and pluckProperty() instead.
+   *
+   * You may write any valid Terraform for each expression, it will result
+   * in `[ for key, val in var.myIteratorSource: <expression> ]`.
+   *
+   * As this returns an IResolvable you might need to wrap the output in
+   * a Token, e.g. `Token.asString`.
```
Comment (mutahhir) [inline packages/cdktf/lib/terraform-iterator.ts:304]: Would be nice to accompany the above terraform example with a CDKTF call showing how to use this API

**Review by mutahhir — APPROVED** [review]
Very nice!
